File: Data Structures & Algorithms/combination-target-sum-ii/submission-8.js
class Solution {
    constructor() {
        this.res = new Set();
    }

    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        this.res.clear();
        candidates.sort((a, b) => a - b);
        this.generateSubsets(candidates, target, 0, [], 0);
        return Array.from(this.res, subset => JSON.parse(subset));
    }

    /**
     * @param {number[]} candidates
     * @param {number} target
     * @param {number} i
     * @param {number[]} cur
     * @param {number} total
     * @return {void}
     */
    generateSubsets(candidates, target, i, cur, total) {
        if (total === target) {
            this.res.add(JSON.stringify([...cur]));
            return;
        }
        if (total > target || i === candidates.length) {
            return;
        }

        cur.push(candidates[i]);
        this.generateSubsets(candidates, target, i + 1, cur, total + candidates[i]);
        cur.pop();

        this.generateSubsets(candidates, target, i + 1, cur, total);
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-8.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<std::vector<int>> res;
        sort(nums.begin(), nums.end());
        dfs(0, {}, 0, nums, target, res);
        return res;
    }

    void dfs(int i, vector<int> cur, int total, vector<int>& nums, int target, vector<vector<int>>& res) {
        if (total == target) {
            res.push_back(cur);
            return;
        }
        
        for (int j = i; j < nums.size(); j++) {
            if (total + nums[j] > target) {
                return;
            }
            cur.push_back(nums[j]);
            dfs(j, cur, total + nums[j], nums, target, res);
            cur.pop_back();
        }
    }
};